compiler/rustc_type_ir/src/sty.rs

```diff
@@ -18,6 +18,34 @@ use self::TyKind::*;
 use rustc_data_structures::stable_hasher::HashStable;
 use rustc_serialize::{Decodable, Decoder, Encodable};
 
+/// Specifies how a trait object is represented.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Encodable, Decodable)]
+pub enum DynKind {
+    /// An unsized `dyn Trait` object
+    Dyn,
+    /// A sized `dyn* Trait` object
+    ///
+    /// These objects are represented as a `(data, vtable)` pair where `data` is a usized value
+    /// (often a pointer to the real object, but not necessarily) and `vtable` is a pointer to
```

Comment:
Who decides the representation, where is that information (about the representation choice) encoded?

In the interpreter, it seems to always be a pointer to the real object (judging from the `cast` code, which is the only new code for this atm).

Reply:
The layout code sets the representation here: https://github.com/rust-lang/rust/pull/101212/files#diff-b0320c2b8868f325d83c027fc5d71732636e9763551e35895488f30fe057c6e9R627-R633

We construct these values here: https://github.com/rust-lang/rust/pull/101212/files#diff-db08b7c8b00530c7183cf2e42f25dc93b02da93fb40edadbd009eee6ad60e3a1R275-R289

Reply:
The first link does not set how things are represented, via pointer or via direct inlining. I was wondering how that is decided.

However the first link does say that the `data` part must be non-null, which should also be documented here.

Not sure where the second link is supposed to go; links into the 'diff' view of a PR are notoriously unrelaible...

Reply:
This should be more stable for the second link: https://github.com/rust-lang/rust/blob/242f9c128ebd5e107fc45ea3725e4e9745ff355b/compiler/rustc_codegen_ssa/src/mir/rvalue.rs#L275...L289

At the moment the data part is made by basically transmuting whatever we are casting into it. The tests so far all work by casting a `usize` into a `dyn*`, for example: [src/test/ui/dyn-star/method.rs](https://github.com/rust-lang/rust/pull/101212/files#diff-aaace7a0cda4100dc786d54d3e80b7945bcb1e0848851b4b16fc152ed4bc9eed).

Longer term we want to do this conversion using a family of lang traits called something like `IntoDynPointer`. I discuss this some in my [blog post on async functions in traits](https://blog.theincredibleholk.org/blog/2022/04/18/how-async-functions-in-traits-could-work-in-rustc/#dynamic-traits-with-dyn-). These would let types override the behavior. By default the implementation would probably be a straight cast for things that are already pointers and `Box::new(self).into_raw()` for things that are not, but we'd like to give flexibility in where async function return futures go.

The non-null requirement in the first link is incorrect, since it's totally reasonable to do `0 as dyn* Debug`. I'll update that.

Reply:
Oh I see, so something somewhere checks that the LHS of an `as dyn*` is ptr-sized?

Reply:
It will in the future, but it doesn't now iirc...